### PR TITLE
feat(fields}: add parser wrapper capable of formulas in formulas [CLK-500345]

### PR DESCRIPTION
**Pull request by mysza**
## Allows for using formulas in formulas in a form of `CUSTOM_FIELD_*` variables.

This is a different API than the current `Parser`, and this functionality is only present when using the `ClickUpParser` object. This class is a wrapper around the underlying parser with some added functionality to make the parsing work, and doing things like controlling maximum nesting level and circular dependencies.

The `ClickUpParser` requires the values of the custom field variables passed to the parser to be of an object type with a shape `{ type: "<field_type>", value: "<field_value>" }` for the "formula" type variables; all other variables can be set either the new way, or the previous way (it's backwards compatible). This is enforced on the `setVariable` function level.

## Testing
- all current tests pass
- integration tests are updated to test all existing behaviour with the `ClickUpParser` instance
- new functionality is covered with unit tests

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -114,8 +114,10 @@ export default class Parser {
         let value = this.getVariable(name);
 
         this.emitter.emit('callVariable', name, (newValue) => {
-            if (newValue !== void 0) {
-                value = newValue;
+            if (newValue !== undefined && newValue !== null && newValue.error) {
+                throw Error(newValue.error);
+            } else if (newValue !== undefined) {
```
Comment (nikitaclicks) [inline src/parser.js:119]: can `newValue` be null?

Comment (mysza) [reply src/parser.js:123]: Yeah it can, but this check was not enough nevertheless - fixed!

File: src/clickup/clickupParser.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+                throw new Error(ERROR_CYCLE);
+            }
+
+            // change context
+            this.evaluationContext.evaluating.add(name);
+            this.evaluationContext.level++;
+
+            // check if we are not exceeding the max levels
+            if (this.evaluationContext.level > this.evaluationContext.maxLevels) {
+                throw new Error(ERROR_LEVEL);
```
Comment (nikitaclicks) [inline src/clickup/clickupParser.ts:55]: Would it be possible to bypass this rule using a conditional formula?

File: test/unit/clickup/customField.test.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+            'variable is not defined',
+            'CUSTOM_FIELD_4',
+            undefined,
+            { name: 'CUSTOM_FIELD_4', type: 'unknown', value: undefined },
+        ],
+        [
+            'variable is null', // null is a valid value
+            'CUSTOM_FIELD_5',
+            null,
+            { name: 'CUSTOM_FIELD_5', type: 'unknown', value: null },
```
Comment (nikitaclicks) [inline test/unit/clickup/customField.test.ts:44]: Wouldn't hurt to include a case with a formula variable

**Review by nikitaclicks — APPROVED** [review]
fantastic work!

File: src/clickup/clickupParser.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    private getCustomFieldVariableValueRetriever(evaluate: (name: FieldName) => FieldValue) {
+        return (name: FieldName, done: (newValue: unknown) => void) => {
+            // check if we are not in a cycle
+            if (this.evaluationContext.evaluating.has(name)) {
+                throw new Error(ERROR_CYCLE);
+            }
+
+            // change context
+            this.evaluationContext.evaluating.add(name);
+            this.evaluationContext.level++;
```
Comment (nikitaclicks) [inline src/clickup/clickupParser.ts:51]: How does this relate to the second PR which also checks nesting level?

File: src/clickup/clickupParser.ts (hunk 2 above)
Comment (mysza) [reply src/clickup/clickupParser.ts:55]: Hm, I think yes if the condition evaluates to non-formula variable... I don't have an idea now how to work around it except of checking the nesting on creation/update.

File: src/clickup/clickupParser.ts (hunk 4 above)
Comment (mysza) [reply src/clickup/clickupParser.ts:51]: The second PR checks the nesting level without evaluation, by just parsing the formula and traversing the variables dependency graph. It is slower, though (still fast, but slower than formula evaluation). Both check for the same thing though.
But I would feel a little unsafe if we didn't check the nesting level on execution :)

File: test/unit/clickup/customField.test.ts (hunk 3 above)
Comment (mysza) [reply test/unit/clickup/customField.test.ts:44]: Added below 👍🏻

**Review by szygendab — APPROVED** [review]

**Pull request merged**